src/main/java/io/odpf/firehose/sink/mongodb/MongoSink.java

```diff
@@ -0,0 +1,99 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static io.odpf.firehose.metrics.Metrics.SINK_MESSAGES_DROP_TOTAL;
+
+/**
+ * MongoDB sink.
+ */
+public class MongoSink extends AbstractSink {
+
+    private final MongoRequestHandler mongoRequestHandler;
+    private final List<WriteModel<Document>> bulkRequest = new ArrayList<>();
+    private final List<String> mongoRetryStatusCodeBlacklist;
+    private final MongoResponseHandler mongoResponseHandler;
+    private final MongoClient mongoClient;
+    private List<Message> messages;
+
+    /**
+     * Instantiates a new Mongo sink.
+     *
+     * @param instrumentation               the instrumentation
+     * @param sinkType                      the sink type
+     * @param mongoCollection               the mongo collection
+     * @param mongoClient                   the mongo client
+     * @param mongoRequestHandler           the mongo request handler
+     * @param mongoRetryStatusCodeBlacklist the mongo retry status code blacklist
+     */
+    public MongoSink(Instrumentation instrumentation, String sinkType, MongoCollection<Document> mongoCollection, MongoClient mongoClient, MongoRequestHandler mongoRequestHandler,
+                     List<String> mongoRetryStatusCodeBlacklist) {
+        super(instrumentation, sinkType);
+        this.mongoRequestHandler = mongoRequestHandler;
+        this.mongoClient = mongoClient;
+        this.mongoRetryStatusCodeBlacklist = mongoRetryStatusCodeBlacklist;
+        this.mongoResponseHandler = new MongoResponseHandler(mongoCollection, instrumentation);
+    }
+
+    @Override
+    protected void prepare(List<Message> messageList) {
+        this.messages = messageList;
+        bulkRequest.clear();
+
+        messages.forEach(message -> bulkRequest.add(mongoRequestHandler.getRequest(message)));
+    }
+
+    @Override
+    protected List<Message> execute() throws Exception {
+        List<BulkWriteError> bulkWriteErrors = processRequest(bulkRequest);
+
+        if (bulkWriteErrors.isEmpty()) {
+            return Collections.emptyList();
+        }
+        getInstrumentation().logWarn("Bulk request failed");
+        handleBulkWriteErrors(bulkWriteErrors);
+
+        return bulkWriteErrors.stream()
+                .filter(bulkWriteError -> !mongoRetryStatusCodeBlacklist.contains(String.valueOf(bulkWriteError.getCode())))
+                .map(bulkWriteError -> messages.get(bulkWriteError.getIndex())).collect(Collectors.toList());
+    }
+
+    @Override
+    public void close() throws IOException {
+        getInstrumentation().logInfo("MongoDB connection closing");
+        this.mongoClient.close();
+    }
+
+    protected List<BulkWriteError> processRequest(List<WriteModel<Document>> bulkRequest) {
+        return mongoResponseHandler.processRequest(bulkRequest);
+    }
+
+    private void handleBulkWriteErrors(List<BulkWriteError> bulkWriteErrors) {
```

Comment:
you can move this method inside ResponseHandler.

Reply:
fixed